src/runtime/type.ts

```diff
@@ -1226,61 +1229,52 @@ export class TypeVariableInfo {
    * Merge a type variable's read subset (upper bound) constraints into this variable.
    * This is used to accumulate read constraints when resolving a handle's type.
    */
-  maybeMergeCanReadSubset(constraint: Type): boolean {
-    if (constraint == null) {
-      return true;
-    }
-
-    if (this.canReadSubset == null) {
-      this.canReadSubset = constraint;
-      return true;
-    }
-
-    if (this.canReadSubset instanceof SlotType && constraint instanceof SlotType) {
-      // TODO: formFactor compatibility, etc.
-      return true;
-    }
-    if (this.canReadSubset instanceof EntityType && constraint instanceof EntityType) {
-      const mergedSchema = Schema.intersect(this.canReadSubset.entitySchema, constraint.entitySchema);
-      if (!mergedSchema) {
-        return false;
-      }
-
-      this.canReadSubset = new EntityType(mergedSchema);
-      return true;
-    }
-    return false;
+  maybeMergeCanReadSubset(constraint: Type, resolveToMaxType: boolean = false): boolean {
+    const {result, success} = this._maybeMerge(
+      this.canReadSubset,
+      constraint,
+      (this._resolveToMaxType  || resolveToMaxType) ? Schema.union : Schema.intersect,
```

Comment:
I don't understand this change to be honest.

If we merge 2 type variables, I understand that merging read constraints should be done with intersection, as you can only read from both what is available in both.

I don't understand why resolveToMaxType would change this?

Reply:
Yeah, I thought this was necessary. However, when I always `intersect`, all the tests still pass. 

I can either revert the changes to this section of code, or keep the refactoring (i.e. `_maybeMerge`) and remove the use of `_resolveToMaxType`. What do you advise?

Reply:
I like the refactoring. It's definitely incorrect to switch to Schema.union if resoveToMaxType is true.